src/transport/stomp/stomp.client.ts

```diff
@@ -0,0 +1,198 @@
+import { StompFrame } from './stomp.frame';
+import { sleep } from '../../utils';
+import { TimeoutError } from '@nestjs/terminus';
+import { WebSocket } from 'ws';
+import { StompFrameException } from './stomp.exceptions';
+
+// https://stomp.github.io/stomp-specification-1.1.html#Overview
+const VERSIONS = '1.0,1.1';
+
+export default class StompClient {
+  private ws: WebSocket;
+
+  opened = false;
+  private connected = false;
+  private counter = 0;
+  private subscriptions: Record<string, (frame: StompFrame) => void> = {};
+
+  public constructor(
+    private url: string,
+    private login: string | null = null,
+    private passcode: string | null = null,
+    private connectCallback: (frame: StompFrame) => void = (
+      frame: StompFrame | boolean,
+    ) => frame,
+    private errorCallback: (frame: StompFrame) => void = (frame) => frame,
+  ) {
+    this.ws = this.createWebSocket(url);
+  }
+
+  private createWebSocket(url) {
+    const ws: WebSocket = new WebSocket(url);
+    ws.on('open', this.onOpen.bind(this));
+    ws.on('message', this.onMessage.bind(this));
+    ws.on('close', this.onClose.bind(this));
+    ws.on('error', this.onError.bind(this));
+    return ws;
+  }
+
+  private async onError(event) {
+    await this.onClose(event);
+  }
+
+  private async transmit(command, headers, body = '') {
+    const msg = StompFrame.marshall(command, headers, body);
+    this.ws.send(msg.toString());
+  }
+
+  private onOpen() {
+    this.opened = true;
+  }
+
+  private async onClose(event) {
+    this.cleanUp();
+    await sleep(10000);
+    try {
+      await this._reconnect(event);
+    } catch (error) {
+      await this.onClose(event);
+    }
+  }
```

Comment:
it's hard to read this code for me. I would suggest putting the logic responsible for reconnecting directly into the reconnect method and call it from the `onError` and `onClose` methods.

Something like this (pseudocode):
```ts
private async onClose(code: number, reason: Buffer) {
    const closeReason = reason.toString();
    this.logger.warn('WS connection is closed', { code, closeReason });

    await this.reconnect();
  }

  private async onError(error: Error) {
    this.logger.warn('WS connection error', error);
    await this.reconnect();
  }

  private async reconnect(timeout = RECONNECT_TIMEOUT) {
    try {
      this.logger.log('Reconnecting to WS');

      await this.disconnect();
      await this.connect();
    } catch (error) {
      this.logger.error(error);

      await sleep(timeout);
      await this.reconnect(timeout);
    }
  }
```